drivers/w5500/w5500.c

```diff
@@ -0,0 +1,468 @@
+        uint8_t tmp[2];
+        read_rx0_buffer(dev, socket0_read_pointer, tmp, 2u);
+        #if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+        packet_size = ((tmp[0] << 8) | tmp[1]) - 2u;
+        #else
+        packet_size = ((tmp[1] << 8) | tmp[0]) - 2u;
+        #endif
+        ret_value = packet_size;
+
+        if (max_len > 0u) { /* max_len > 0u: Client wants to read or drop the packet. */
+            if (packet_size > max_len) {
+                buf = NULL; /* Drop the packet. */
+                ret_value = -ENOBUFS;
+            }
+
+            socket0_read_pointer += 2u;  /* Add the 2 bytes of the packet_size. */
+            if (buf != NULL) {
+                read_rx0_buffer(dev, socket0_read_pointer, (uint8_t *)buf, packet_size);
+                DEBUG("[w5500] receive: got packet of %i byte (at 0x%04x)\n", packet_size,
+                      (int)socket0_read_pointer);
+            }
+            else {
+                DEBUG("[w5500] receive: drop packet of %i byte (at 0x%04x)\n", packet_size,
+                      (int)socket0_read_pointer);
+            }
+            socket0_read_pointer += packet_size;
+        }
+
+        /* Update read pointer. */
+        waddr(dev, REG_S0_RX_RD0, REG_S0_RX_RD1, socket0_read_pointer);
+        wreg(dev, REG_S0_CR, CR_RECV);
+        while (rreg(dev, REG_S0_CR)) {}
+    }
+
+    spi_release(dev->p.spi);
+
+    return ret_value;
+}
+
+static void isr(netdev_t *netdev)
+{
+    w5500_t *dev = (w5500_t *)netdev;
+
+    spi_acquire(dev->p.spi, dev->p.cs, SPI_CONF, dev->p.clk);
+    /* Get phy status register. */
+    uint8_t phy_status = rreg(dev, REG_PHYCFGR);
+    /* Get socket 0 interrupt register. */
+    uint8_t socket0_interrupt_status = rreg(dev, REG_S0_IR);
+
+    /* Clear interrupts. */
+    wreg(dev, REG_S0_IR, socket0_interrupt_status);
+    spi_release(dev->p.spi);
+
+    if ((phy_status & PHY_LINK_UP) && (dev->link_up == false)) {
+        dev->link_up = true;
+        DEBUG("%s: link is up: PHYCFGR = 0x%02x\n", __func__, phy_status);
+        netdev->event_callback(&dev->netdev, NETDEV_EVENT_LINK_UP);
+    }
+    else if (!(phy_status & PHY_LINK_UP) && (dev->link_up == true)) {
+        dev->link_up = false;
+        netdev->event_callback(&dev->netdev, NETDEV_EVENT_LINK_DOWN);
+        DEBUG("%s: link is down: PHYCFGR = 0x%02x\n", __func__, phy_status);
+    }
+
+    while (socket0_interrupt_status & (IR_RECV | IR_SEND_OK)) {
+        if (socket0_interrupt_status & IR_RECV) {
+            DEBUG("[w5500] netdev RX complete\n");
+            netdev->event_callback(netdev, NETDEV_EVENT_RX_COMPLETE);
+        }
+        if (socket0_interrupt_status & IR_SEND_OK) {
+            DEBUG("[w5500] netdev TX complete\n");
+            dev->frame_size_sent = dev->frame_size_to_be_send;
+            dev->frame_size_to_be_send = 0u;
+            netdev->event_callback(netdev, NETDEV_EVENT_TX_COMPLETE);
+        }
+
+        /* Check interrupt status again. */
```

Comment:
Is this only needed in polling mode? 
Mind you that `spi_acquire()` can be surprisingly expensive.

Reply:
This is needed in both modes, because the interrupt status has to be read from the device via SPI.

The reason of aquiring and releasing the SPI bus in isr():
When NETDEV_EVENT_RX_COMPLETE is send up into the gnrc stack the function _event_cb in gnrc_netif.c will call netif->ops->recv(netif) [line 2086], there is no decoupling with a message queue. Following the callstack recv(netif) will call _recv in gnrc_netif_raw [line 53 and later again] which will call the function receive() in w5500.c. This receive function will call spi_aquire to get access to the spi-bus. So the spi bus must have been released before. And re-aquired if the isr function has to check again the interrupt status.

While I wrote this description here I thought I might keep the bus aquired at the beginning and release it at the end of the isr function. But that would mean that the implementation expects the gnrc stack to call the receive-path in the same context. I am not so sure if that is a good idea.